server/src/test/java/org/elasticsearch/index/shard/RefreshListenersTests.java

```diff
@@ -352,43 +433,85 @@ public void testLotsOfThreads() throws Exception {
             assertEquals(0, listeners.pendingCount());
 
             try (Releasable releaseable2 = listeners.forceRefreshes()) {
-                listener = new DummyRefreshListener();
+                listener = new TestLocationListener();
                 assertTrue(listeners.addOrNotify(index("1").getTranslogLocation(), listener));
                 assertTrue(listener.forcedRefresh.get());
                 listener.assertNoError();
+                seqNoListener = new TestSeqNoListener();
+                assertTrue(listeners.addOrNotify(index("1").getSeqNo(), seqNoListener));
+                assertTrue(seqNoListener.isDone.get());
                 assertEquals(0, listeners.pendingCount());
             }
 
-            listener = new DummyRefreshListener();
+            listener = new TestLocationListener();
             assertTrue(listeners.addOrNotify(index("1").getTranslogLocation(), listener));
             assertTrue(listener.forcedRefresh.get());
             listener.assertNoError();
+            seqNoListener = new TestSeqNoListener();
+            assertTrue(listeners.addOrNotify(index("1").getSeqNo(), seqNoListener));
+            assertTrue(seqNoListener.isDone.get());
             assertEquals(0, listeners.pendingCount());
         }
 
-        assertFalse(listeners.addOrNotify(index("1").getTranslogLocation(), new DummyRefreshListener()));
+        assertFalse(listeners.addOrNotify(index("1").getTranslogLocation(), new TestLocationListener()));
+        assertFalse(listeners.addOrNotify(index("1").getSeqNo(), new TestSeqNoListener()));
+        assertEquals(2, listeners.pendingCount());
+    }
+
+    public void testSequenceNumberMustBeIssued() throws Exception {
+        assertEquals(0, listeners.pendingCount());
+        TestSeqNoListener seqNoListener = new TestSeqNoListener();
+        long issued = index("1").getSeqNo();
+        assertTrue(listeners.addOrNotify(issued + 1, seqNoListener));
+        assertThat(seqNoListener.error, instanceOf(IllegalArgumentException.class));
+        String message = "Cannot wait for unissued seqNo checkpoint [wait_for_checkpoint="
+            + (issued + 1) + ", max_issued_seqNo=" + issued + "]";
+        assertThat(seqNoListener.error.getMessage(), equalTo(message));
+    }
+
+    public void testForcedRefreshDoesNotCompleteUnRefreshedCheckpoint() throws Exception {
+        assertEquals(0, listeners.pendingCount());
+
+        Engine.IndexResult index = index("1");
+        TestSeqNoListener seqNoListener = new TestSeqNoListener();
+        for (int i = 0; i < maxListeners; ++i) {
+            listeners.addOrNotify(index.getSeqNo(), seqNoListener);
+        }
+
+        long mockSeqNo = index.getSeqNo() + 1;
+        listeners.setMaxIssuedSeqNoSupplier(() -> mockSeqNo);
+        TestSeqNoListener forcingListener = new TestSeqNoListener();
+        listeners.addOrNotify(mockSeqNo, forcingListener);
+        assertFalse(forcingListener.isDone.get());
```

Comment:
Can we add verification that the `forcingListener` is completed once the processed seq-no advances?